core/src/jmh/java/com/uber/m3/tally/AbstractReporterBenchmark.java

```diff
@@ -0,0 +1,140 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.uber.m3.tally;
+
+import com.uber.m3.util.Duration;
+import com.uber.m3.util.ImmutableMap;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+
+// TODO: 31/01/2021 add descrition and readme how to use.
+@State(Scope.Benchmark)
+public abstract class AbstractReporterBenchmark {
+    private static final ImmutableMap<String, String> DEFAULT_TAGS = new ImmutableMap.Builder<String, String>(5)
+            .put("tag1", "test1")
+            .put("tag2", "test2")
+            .put("tag3", "test3")
+            .put("tag4", "test4")
+            .put("env", "test")
+            .build();
+
+    private static final DurationBuckets DURATION_EXPONENTIAL_BUCKETS = DurationBuckets.linear(
+            Duration.ofMillis(1), Duration.ofMillis(10), 128
+    );
+
+    private static final ValueBuckets VALUE_EXPONENTIAL_BUCKETS = ValueBuckets.linear(
+            0.1, 100d, 128
+    );
+
+    private static final String COUNTER_NAME = "counter";
+    private static final String GAUGE_NAME = "gauge";
+    private static final String TIMER_NAME = "timer";
+    private static final String HISTOGRAM_DURATION_NAME = "histogram_duration";
+    private static final String HISTOGRAM_VALUE_NAME = "histogram_value";
+
+    private StatsReporter reporter;
+
+    @Benchmark
+    public void reportCounterBenchmark(IncrementalValueState value) {
+        reporter.reportCounter(COUNTER_NAME, DEFAULT_TAGS, value.incrementAndGet());
+    }
+
+    @Benchmark
+    public void reportGaugeBenchmark(IncrementalValueState value) {
+        reporter.reportCounter(COUNTER_NAME, DEFAULT_TAGS, value.incrementAndGet());
+    }
+
+    @Benchmark
+    public void reportTimerBenchmark(IncrementalValueState value) {
+        reporter.reportTimer(TIMER_NAME, DEFAULT_TAGS, Duration.ofSeconds(value.incrementAndGet()));
+    }
+
+    @Benchmark
+    public void reportHistogramDurationSamplesBenchmark(IncrementalValueState value) {
+        reporter.reportHistogramDurationSamples(
```

Comment:
Sanity checking: I'm assuming we can't get deadcode eliminated here, because this is all side effect based--sound right?

Reply:
yep, maybe we should add some ignore annotations so that IDE-s don't complain about it

Reply:
Oh do they complain about it? I wouldn't assume so but not sure. Anyhow thanks for confirming!